Review pip-only-in-virtualenv (#3)
geofft/peps: pep-9999.rst

PEP: 9999
Title: Graceful cooperation between external and Python package managers

> Author: Geoffrey Thomas <geofft@ldpreload.com>, Matthias Klose <doko@ubuntu.com>, Filipe Laíns <lains@riseup.net>, Donald Stufft <donald@python.org>, Tzu-Ping Chung <uranusjr@gmail.com>, Stefano Rivera <stefanor@debian.org>, Elana Hashman <ehashman@debian.org>, Pradyun Gedam <mail@pradyunsg.me>

Reviewer: There's a very specific format to write multiple authors (see e.g. [PEP 622](https://github.com/python/peps/blob/master/pep-0622.rst)). I'm not sure if it's syntactically requried or just an agreed style.

Discussions-To: TODO discourse
Status: Draft
Type: Informational
Content-Type: text/x-rst
Created: 2021-05-18
Post-History:

Abstract
========

This PEP proposes a mechanism to indicate that a certain
Python interpreter's package installation is managed by
external means (such as an OS package manager) and that
Python-specific package management tools like pip should, by
default, neither install nor remove packages in an
interpreter-wide context, guiding the end user towards using

> a virtual environment.

Reviewer 2: This wording is really difficult to grok. I'd suggest we rewrite this paragraph. Perhaps,

> There are contexts (such as an OS package manager) where Python-specific package management tools like `pip` should neither install nor remove packages by default, and instead guide the end user towards using a virtual environment. This PEP proposes a mechanism for Python installations to communicate this.

Author: What do you think about replacing this paragraph with the following three?

> A long-standing practical problem for Python users has been  conflicts between OS package managers and Python-specific package management tools like pip. These conflicts include both Python-level API incompatibilities and conflicts over file ownership.
>
> Historically, Python-specific package management tools have defaulted to installing packages into an implicit global context. With the standardization and popularity of virtual environments, a better solution for most (but not all) use cases is to use Python-specific package management tools only within a virtual environment.
>
> This PEP proposes a mechanism for a Python installation to communicate to tools like pip that its global package
installation context is managed by some means external to Python, such as an OS package manager. It specifies that  Python-specific package management tools should neither install nor remove packages into the interpreter's global context, by default, and should instead guide the end user towards using a virtual environment.

Even with the last paragraph, it comes out to exactly 200 words, which is the suggestion from PEP 1 :)

Author: (pushed that change)

It also standardizes an interpretation of the ``sysconfig``
schemes so that, if a Python-specific package manager is

> specifically instructed to install a package in an
> interpreter-wide context, it can do so in a manner that
> avoids conflicting with the external package manager and

Reviewer 2: ```suggestion
instructed to install a package in an
interpreter-wide context, it can do so in a manner that
will avoid conflicting with the external package manager and
```

Author: I want to convey in some fashion that while `sudo pip install foo` technically counts as being instructed to install into /usr/local, the thing in the previous paragraph means it's going to refuse.

Maybe replace "is instructed to install" with "is about to install", which shifts from what it's being asked to do to what it has decided to actually do?

Reviewer 2: "specifically" was superfluous, the other was a grammar nit

reduces the risk of breaking software shipped by the
external package manager.

Motivation
==========

Thanks to Python's immense popularity, software
distributions (by which we mean Linux and other OS
distributions as well as overlay distributions like Homebrew
and MacPorts) generally ship Python for two purposes: as a
software package to be used in its own right by end users,

> and as an implementation language for other software in the

Reviewer 2: ```suggestion
and as a language dependency for other software in the
```

distribution.

> For example, many distributions ship a ``/usr/bin/python3``
> command which is both the ``python3`` command available to
> end users as well as the ``#!/usr/bin/python3`` shebang for
> Python-language software. Because there are no official

Reviewer 2: ```suggestion
For example, many distributions ship a ``/usr/bin/python3``
binary or script, which provides the ``python3`` command available to
end users as well as the ``#!/usr/bin/python3`` shebang for
Python-language software included in the distribution. Because there are no official
```

Author: (wait who ships /usr/bin/python3 as a script?)

Reviewer 2: Idk about Python3 specifically but a ton of binary entrypoints in distros are wrapper scripts.

> binary releases of Python for Linux/UNIX, almost all Python
> end users on these OSes use the Python interpreter built and
> shipped with their distribution.

Reviewer 2: ```suggestion
binary releases of Python for Linux/UNIX, almost all Python
end users on these operating systems use the Python interpreter built and
shipped with their distribution in some way, even if they compile their
own Python binaries.
```

Author: Hm I think this changes the focus of the statement away from "most people do have a dual-use Python" to "most people use the distro Python interpreter in at least one of the two ways above," but if they only use it in one way then the problem kind of goes away. What's the change you're trying to convey here?

Reviewer 2: A lot of people will claim it's bad practice to have a dual-use Python, when in reality they still use the system Python in a non-trivial way, whether or not they realize it.

Author: How about s/almost all/most/, and then a new paragraph after the mention of `pip install --user`, something like

> Even for Python end users who avoid using the system-provided Python interpreter, such as by compiling their own from source, packages installed by either interpreter via `sudo pip install` or `pip install --user` will be visible on the `sys.path` of both interpreters. (As discussed later in this document, some distros configure their interpreter to use a non-default location for `sudo pip install` to address this use case, but no known distro customizes the `pip install --user` location / the user site-packages directory.)

> These are typically the same command, which means that if an

Reviewer 2: ```suggestion
The `python3` executable available to the users of the distribution 
and the `python3` executable available as a dependency for other 
software in the distribution are typically the same binary. This means that if an
```

end user installs a Python package using a tool like ``pip``

> outside the context of a virtual environment, the new
> package is visible to Python-language software shipped by

Reviewer 2: ```suggestion
outside the context of a virtual environment, that
package is visible to Python-language software shipped by
```

the distribution. If the new package is not

> backwards-compatible, it may break software shipped by the

Reviewer 2: I'm not sure what "backwards-compatible" means here. More importantly, the process by which that new package was installed may break the distribution when it resolves and updates dependencies.

Author: I mean regular old API backwards compatibility - in particular, the `python-requests` in Debian 7 (some 0.x) had `.json` as a property but requests 1.x just uses `.json()` as a normal function, so if you `sudo pip install -U requests`, you break code written against the OS version of requests.

I think you're referencing another and also interesting problem about breaking the distribution... can we get into a state where `pip` leaves dpkg unable to deal with what's on the filesystem? If so, let's mention/document it. (I was thinking about the case where Python-language code such as `dnf` breaks.)

Author: Oh I realized there was no antecedent to _what_ the new package is supposed to be ackwwards-compatible with. I expanded this sentence, it should make more sense now.

distribution.

> This problem is particularly extreme because some
> distributions have package-management tools that are
> themselves written in Python. For example, it's possible to

Reviewer 2: ```suggestion
This may pose a critical problem for the integrity of distributions,
which often have package-management tools that are
themselves written in Python. For example, it's possible to
```

best way to install Python libraries or applications (when
not using a distribution package) is to use a virtual
environment. This approach was popularized by the PyPA
`virtualenv`_ project, and a simple version of that approach
is now available in the Python standard library as ``venv``.
Installing a Python package into a virtualenv prevents it
from being visible to the unqualified ``/usr/bin/python3``
interpreter and prevents breaking system software.

.. _virtualenv: https://virtualenv.pypa.io/en/latest/

In some cases, however, it's useful and intentional to
install a Python library from outside of the distribution
that influences the behavior of distribution-shipped
commands. This is common in the case of software like Sphinx
or Ansible which have a mechanism for writing
Python-language extensions. A user may want to use their
distribution's version of the base software (for reasons of
paid support or security updates) but install a small
extension from PyPI, and they'd want that extension to be
importable by the software in their base system.

While this continues to carry the risk of installing a newer
version of a dependency than the operating system expects or

> otherwise negatively changing the behavior of an

Reviewer 2: ```suggestion
otherwise negatively affecting the behavior of an
```

directories, one for its own managed packages, and one for
unmanaged packages installed by the end user, and ensure
that installing unmanaged packages will not delete (or
overwrite) files owned by the external package manager.

Rationale
=========

As described in detail in the next section, the first
behavior change involves creating a marker file named
``EXTERNALLY-MANAGED``, whose presence indicates that
non-virtual-environment package installations are managed by
some means external to Python, such as a distribution's
package manager. This file is specified
to live in the ``stdlib`` directory in the default
``sysconfig`` scheme, which marks the interpreter /
installation as a whole, not a particular location on
``sys.path``. The reason for this is that, as identified
above, there are two related problems that risk breaking an
externally-managed Python: you can install an incompatible
new version of a package system-wide (e.g., with ``sudo pip
install``), and you can install one locally, but in a
location that is loaded by the standard Python command
(e.g., with ``pip install --user``). If the marker file were
in the system-wide ``site-packages`` directory, it would not

> clearly apply to the second case. The `Rejected Ideas`_

Reviewer 2: ```suggestion
clearly apply to the second case. The `Alternatives`_
```

might be a bit softer, particularly if anyone ends up arguing that we go with those solutions... :)

Author: "Rejected Ideas" is the name of a standard PEP section, see https://www.python.org/dev/peps/pep-0001/#what-belongs-in-a-successful-pep . I'm not opposed to changing it, but that's why I defaulted to it.

Reviewer 2: That's a shame. I wouldn't want changing this to hurt the PEP's chances.

Author: Thinking harder, since it's got the "System Python" subsection, we definitely don't want to call it "Rejected Ideas" because we're emphatically taking no position on it. So, amendment is friendly.

section has further discussion of possible locations.

The second behavior change takes advantage of the existing
``sysconfig`` setup in distributions that have already
encountered this class of problem, and specifically
addresses the problem of a Python-specific package manager
deleting or overwriting files that are owned by an external
package manager.

> The changed behavior in this PEP is intended to "do the

Reviewer 2: This section needs a header. Perhaps `Use cases`? It's no longer discussing rationale from here down.

Author: I'll do a subsection header - the rationale for the specific choices is that, in the following use cases, it DTRTs.

right thing" for as many use cases as possible. For the
following representative use cases, we analyze the two
behaviors that could be changed by this PEP: can you run a
Python-specific installer tool like ``pip install`` by
default, and if you do run such a tool, should it be willing
to delete packages shipped by the external package manager?
For simplicity, this section refers to pip's behavior,
although it applies equivalently to any other
Python-specific package management tool.

1. A standard unpatched CPython, without any special
   configuration of or patches to ``sysconfig`` and without
   a marker file. The behavior does not change:

   * Can you ``pip install`` without an override? Currently
     yes; after this PEP yes.

   * Can it delete externally-installed packages? Currently
     yes; after this PEP yes. Because the ``sysconfig``
     scheme used by any externally-installed packages is the
     same scheme used by ``pip install`` (which is to say,
     probably any externally-installed packages were

>      installed by pip too), there is no distinction.

Reviewer 2: Is this desirable? I can't say I'd want to compile CPython for my Debian machine and then have break my system Python. Why wouldn't it respect the distribution hints?

Author: If you install upstream CPython and `sudo /usr/local/bin/python -m pip uninstall` something, it's not going to see distribution-installed things in `/usr/lib/python3/dist-packages` at all, i.e., "externally-installed packages" in this context doesn't mean distribution-installed packages because there is no relevant distribution. For the same reason it doesn't respect the distribution EXTERNALLY-MANAGED file because it doesn't see that file at all.

(If you install upstream CPython with, idk, `./configure -prefix=/usr` then it will see the EXTERNALLY-MANAGED file and the answer to at least the first question is yes->no, but also if you install upstream CPython into /usr, you're doing something extremely strange.)

Should I rephrase this to say something like "irrelevant because there is no such thing as externally-installed packages in this context"?

Author: Rephrased this a bit, see if it's clearer.

     no, because pip has a special case for running in a
     virtual environment and deleting packages outside it.
     After this PEP, still no, in a more general way,
     because system site packages will be outside any of the
     ``sysconfig`` schemes for the environment.

5. A distribution Python when used in a single-application
   container image (e.g., a Docker container). In this use
   case, the risk of breaking system software is lower,
   since generally only a single application runs in the
   container, and the impact is lower, since you can rebuild
   the container and you don't have to struggle to recover a
   running machine. There are also a large number of
   existing ``Dockerfile``s with an unqualified ``RUN pip
   install ...`` statement, etc., and it would be good not
   to break those.  So, builders of base container images
   may want to ensure that the marker file is not present,
   even if the underlying OS ships one by default. In such a
   case:

   * Can you ``pip install`` without an override? Currently
     yes; after this PEP yes.

   * Can it delete externally-installed packages? Currently

>      yes; after this PEP no. This is a behavior change which

Reviewer 2: Why is this particular case yes/no? I don't understand how this is different than the regular distro case.

Author: Not sure I follow - it is not different from the regular distro case (case 2), which is also yes->no.

The behavior change between case 2 and case 4 is in the "Can you pip install" question, which is yes->no for case 2 and yes->yes for case 4.

Reviewer 2: Can we just turn this section into like... a giant table?

Author: That's what I was originally going to do but there's a lot of explanation on each one. Maybe table up front, rationale below?

Reviewer: A table with numbers attached to each cell and elaborate more in numbered points below the table could work well.

Author: I added a table, but I think I managed not to actually reduce the explanation below by very much. :) Still, it should be more readable, take a look?

   in your code, write a downstream ``Dockerfile`` that
   applies OS upgrades, etc., any packages you've installed
   cannot impact the OS.

6. A Conda environment, where a non-``conda`` tool like pip
   is being used to remove packages. In this context, Conda
   acts as the external package manager and pip as the
   Python-specific one. This is currently supported by Conda
   as a way to install software not available in the Conda
   repositories.

   In some sense, this is similar to the first case, since
   Conda provides its own installation of the Python
   interpreter.

   * Can you ``pip install`` without an override? Currently
     yes; after this PEP yes, since Conda would not want to
     add the marker to prevent pip from running.

   * Can it delete externally-installed packages? Currently
     yes; after this PEP yes, unless Conda changes their
     ``sysconfig`` schemes to use separate ones for
     pip-installed and Conda-installed software.  Currently
     those are the same scheme, but it may be a good idea
     for Conda to change these, because if pip removes a
     package, it doesn't inform Conda of that.

> 7. A build of Python as part of an integrated monorepo with
>    its own dependency tooling, such as a private monorepo at

Reviewer 2: ```suggestion
7. A build of Python as part of an integrated distribution with
   its own dependency tooling, such as a private base platform at
```

Author: See email, there's a different rephrasing which you may or may not prefer for this case.

   a software development company or an integrated package
   manager like Nixpkgs_.  In these cases, the authors of

>    the monorepo may want to encourage use of the monorepo's

Reviewer 2: ```suggestion
   the distribution may want to encourage use of the distribution's
```

   own mechanisms for importing Python packages, for the
   same reasons as a software distribution would. They may
   also have custom instructions for how to properly set up
   a virtual environment.

   .. _Nixpkgs: https://github.com/NixOS/nixpkgs

   * Can you ``pip install`` without an override? Currently
     yes (at least ``pip install --user``); after this PEP
     no, because they would include a marker file.

   * Can it delete externally-installed packages? Currently
     possibly yes, depending on the details of how the

>      monorepo finds dependencies and whether it's writable;

Reviewer 2: ```suggestion
     distribution finds dependencies and whether it's writable;
```

   library - e.g., building a ``python-xyz`` RPM by using
   ``pip install xyz`` inside the RPM spec file.

   * Can you ``pip install`` without an override? Currently
     yes; after this PEP, the build environment would need
     to find some way to suppress the marker file to allow
     ``pip install`` to work, but yes, provided it does
     that.

     See the `Recommendations for distributions`_ section
     for more discussion on how to implement this.

   * Can it delete externally-installed packages? Currently
     yes (assuming the distribution doesn't have an
     equivalent to the Debian patch mentioned earlier);
     after this PEP no. This change is fine because a
     package build process should not (and generally cannot)
     include instructions to delete some other files on the
     system; it can only package up its own files.

9. A distribution Python used with ``PYTHONHOME`` to set up
   an alternative Python environment (as opposed to a
   virtual environment).

   If the ``PYTHONHOME`` is copied directly from the
   distribution Python (e.g., ``cp -a /usr/lib/python3.x
   pyhome/lib``), with no modifications, then the behavior

>    is just like a distribution Python:

Reviewer 2: What number is that scenario?

Author: case 2. will add this in a parenthetical

     the marker file.

   * Can it delete externally-installed packages? Currently
     yes; after this PEP no, assuming the distribution used
     separate ``sysconfig`` schemes.

   This is a behavior change, but it seems to be defensible,
   in that if your ``PYTHONHOME`` is a straight copy of the
   distribution's Python, it should behave like the
   distribution's Python.

   If the ``PYTHONHOME`` is separate from the distribution,
   such as a copy of the standard library from an unmodified
   upstream Python (but used with a compatible interpreter
   from the distribution), then the behavior is just like an
   unmodified upstream Python and does not change:

   * Can you ``pip install`` without an override? Currently
     yes; after this PEP yes, because there is no marker
     file.

   * Can it delete externally-installed packages? Currently
     yes; after this PEP yes, because the distribution's
     changes to ``sysconfig`` won't be visible, and there
     will only be one shared scheme.

Specification
=============

Marking an interpreter as using an external package manager
-----------------------------------------------------------

> Before a Python package installer (that is, a tool such as

Reviewer 2: ```suggestion
Before a Python-specific package installer (that is, a tool such as
```

users can install whatever they like (as they can today)
without having to manually override this rule.

Writing to only the target ``sysconfig`` scheme
-----------------------------------------------

Usually, a Python package installer installs to directories
in a scheme returned by the ``sysconfig`` standard library
package.  Ordinarily, this is the scheme returned by
``sysconfig.get_default_scheme()``, but based on
configuration (e.g. ``pip install --user``), it may use a
different scheme.

Whenever the installer is installing to a ``sysconfig``
scheme, this PEP specifies that the installer should never
modify or delete files outside of that scheme. For instance,
if it's upgrading a package, and the package is already
installed in a directory outside that scheme (perhaps in a
directory from another scheme), it should leave the existing
files alone.

If the installer does end up shadowing an existing
installation during an upgrade, we recommend that it
produces a warning at the end of its run.

If the installer is installing to a location outside of a
``sysconfig`` scheme (e.g., ``pip install --target``), then
this subsection does not apply.

Recommendations for distributions

> =================================

Reviewer 2: This section would be more readible if we used subheadings rather than bullet points.

> This section is non-normative. It provides advice we think
> is a good idea for distributions to follow unless they have
> a specific reason otherwise.

Reviewer 2: ```suggestion
This section is non-normative. It provides best practices
we believe distributions should follow unless they have
a specific reason otherwise.
```

  error message indicating both how to install system-wide
  packages via the distribution's package manager and how to
  set up a virtual environment. If your distribution is
  often used by users in a state where the ``python3``
  command is available (and especially where ``pip`` or
  ``get-pip`` is available) but ``python3 -m venv`` does not
  work, the message should indicate clearly how to make
  ``python3 -m venv`` work properly.

* Consider packaging pipx_, a tool for installing
  Python-language applications, and suggesting it in the
  error. It automatically creates a virtual environment for
  that application alone, which is a much better default for
  end users who want to install some Python-language
  software (which isn't available in the distribution) but
  are not themselves Python users. Packaging pipx in the
  distribution avoids the irony of instructing users to
  ``pip install --user --break-system-packages pipx`` to
  _avoid_ breaking system packages.  Consider arranging
  things so your distribution's package / environment for
  Python for end users (e.g., ``python3`` on Fedora or
  ``python3-full`` on Debian) depends on pipx.

> .. _pipx: https://github.com/pipxproject/pipx

Reviewer 2: Soon to be a pypa namespaced project!

Author: Yup, I figure that will happen before the PEP hits discuss and I can just change it - if you think it'll take longer I can add some text to that effect.

Reviewer 3: It happened: https://github.com/pypa/pipx

pip install`` or ``pip install --user`` and add files to
``sys.path`` _for security reasons_, that needs to be
implemented either via access control rules on what files
the user can write to or an explicitly secured ``sys.path``
for the program in question. Neither of the mechanisms in
this PEP should be interpreted as a way to address such a
scenario.

For those reasons, an attempted install with a marker file
present is not a security incident, and there is no need to
raise an auditing event for it. If the calling user
legitimately has access to ``sudo pip install`` or ``pip
install --user``, they can accomplish the same installation
entirely outside of Python; if they do not legitimately have
such access, that's a problem outside the scope of this PEP.

The marker file itself is located in the standard library
directory, which is a trusted location (i.e., anyone who can
write to the marker file used by a particular installer
could, presumably, run arbitrary code inside the installer).
Therefore, there is generally no need to filter out terminal
escape sequences or other potentially-malicious content in
the error message.

> Rejected Ideas

Reviewer 2: ```suggestion
Alternatives
```

==============

> There are a number of similar-sounding proposals that this
> PEP rejects or defers, largely to preserve the behavior in

Reviewer 2: ```suggestion
There are a number of similar proposals we considered that this
PEP rejects or defers, largely to preserve the behavior in
```

Type: Informational
Content-Type: text/x-rst
Created: 2021-05-18
Post-History:

Abstract
========

This PEP proposes a mechanism to indicate that a certain
Python interpreter's package installation is managed by
external means (such as an OS package manager) and that
Python-specific package management tools like pip should, by
default, neither install nor remove packages in an
interpreter-wide context, guiding the end user towards using
a virtual environment.

It also standardizes an interpretation of the ``sysconfig``
schemes so that, if a Python-specific package manager is
specifically instructed to install a package in an
interpreter-wide context, it can do so in a manner that
avoids conflicting with the external package manager and
reduces the risk of breaking software shipped by the
external package manager.

Motivation
==========

Thanks to Python's immense popularity, software
distributions (by which we mean Linux and other OS
distributions as well as overlay distributions like Homebrew
and MacPorts) generally ship Python for two purposes: as a
software package to be used in its own right by end users,
and as a language dependency for other software in the
distribution.

> For example, many distributions ship a ``/usr/bin/python3``

Reviewer: <em><sup>[citation needed] [who?]</em></sup>

I would name some specific distributions that do this instead. That is enough for motivation, and avoids some obvious questions.

    requires multiple entries on ``sys.path``: if package A
    2.0 installs module ``a.py`` in one ``sys.path`` entry,
    and package A 1.0 installs module ``a.py`` in a later
    ``sys.path`` entry, then ``import a`` returns the module
    from the former, and we say that A 2.0 shadows A 1.0.

Motivation
==========

Thanks to Python's immense popularity, software
distributions (by which we mean Linux and other OS
distributions as well as overlay distributions like Homebrew
and MacPorts) generally ship Python for two purposes: as a
software package to be used in its own right by end users,
and as a language dependency for other software in the
distribution.

For example, Fedora and Debian (and their downstream
distributions, as well as many others) ship a
``/usr/bin/python3`` binary which provides the ``python3``
command available to end users as well as the
``#!/usr/bin/python3`` shebang for Python-language software
included in the distribution. Because there are no official
binary releases of Python for Linux/UNIX, almost all Python
end users on these OSes use the Python interpreter built and
shipped with their distribution.

> The `python3` executable available to the users of the distribution 

Reviewer: Needs double backticks instead (many occurrences below).

interpreter compiled from upstream sources will not be so
marked, and so running ``pip install`` with a self-compiled
interpreter, or with a distribution that has not explicitly
marked its interpreter, will work as it always has worked.

Second, it sets the rule that when installing packages to an
interpreter's global context (either to an unmarked
interpreter, or if overriding the marking),
**Python-specific package managers should modify or delete
files only within the directories of the sysconfig
scheme in which they would create files**. This permits a
distributor of a Python interpreter to set up two
directories, one for its own managed packages, and one for
unmanaged packages installed by the end user, and ensure
that installing unmanaged packages will not delete (or
overwrite) files owned by the external package manager.

Rationale
=========

As described in detail in the next section, the first
behavior change involves creating a marker file named
``EXTERNALLY-MANAGED``, whose presence indicates that
non-virtual-environment package installations are managed by
some means external to Python, such as a distribution's
package manager. This file is specified

> to live in the ``stdlib`` directory in the default

Reviewer 4: Why not `platlib` or `purelib`, which will automatically handle the venv case?

Reviewer: Because we still want to block `--user` installs in a virtual environment. pip currently blocks `--user` in a virtual environment, but other installers may not implement the same. Another consideration is that files in `platlib` and `purelib` could be overriden by a package, which is problematic. Finally, this makes scheme-switching more complicated because each scheme has different `platlib` and `purelib` values (should a tool respect the file in the *target* scheme’s `purelib` or the default scheme’s?) while the `stdlib` is always the same.

Reviewer 4: I didn't see anything here that would prevent you from keeping the venv case anyway, or any reason to make that universal (what if you pipx run a tool for managing packages in other environments?).

Presumably even in `stdlib` a package could bring its own copy of the file and start enforcing restrictions. That should probably just be a "don't do that" scenario, since I can actually think of valid reasons to do it (not on PyPI, of course).

I thought it was explicit that you always rely on the default scheme? But I also don't see why you wouldn't look at the directory that you're about to install into and decide then. That's going to make the most sense for everyone involved. (Merging `purelib` and `platlib` I assume is more complex, but then I have to assume that installers currently have a way to handle this - it's not an issue on Windows, so I've never actually had to deeply understand the practical difference).

Reviewer: > Presumably even in `stdlib` a package could bring its own copy of the file and start enforcing restrictions. That should probably just be a "don't do that" scenario, since I can actually think of valid reasons to do it (not on PyPI, of course).

I think that’s a different preference then. From the discussions I was involved in during the PyCon US sprint, it was stated as an explicit goal that other authors *don’t* want this to be possible (I don’t have strong preferences personally).

Reviewer 4: Then either installers have to filter the files they're willing to extract (which I think is fine), or the setting gets moved out of the install directory and into some other configuration location (which I also think is fine). Otherwise there's no way currently to stop a rogue package from adding the file (though as I say, I think that's fine too, and we just file bugs and aggressively fork anyone who tries to do it on PyPI :-) ). Nobody is _removing_ or replacing an existing file without unsupported hacking, of course.

It still seems more consistent with files like `.htaccess` and `.gitignore` to have this configuration affect the directory that it's in. The rejected idea for this presumes that the marker can be found in any `sys.path` entry and then apply to the installation as a whole, but it's not clear whether it was considered to have the marker be discovered (either anywhere, or only in the installers intended target directory) and specify either which schemes/paths may (or may not) be used.

Author: Because if you're installing to the user scheme (with e.g. `pip install --user`), then `platlib` and `purelib` point to your home directory, but we still want to prevent people from breaking system applications by installing things there (by default).

package manager.

Use cases
---------

The changed behavior in this PEP is intended to "do the
right thing" for as many use cases as possible. For the
following representative use cases, we analyze the two
behaviors that could be changed by this PEP: can you run a
Python-specific installer tool like ``pip install`` by
default, and if you do run such a tool, should it be willing
to delete packages shipped by the external package manager?
For simplicity, this section refers to pip's behavior,
although it applies equivalently to any other
Python-specific package management tool.

1. A standard unpatched CPython, without any special
   configuration of or patches to ``sysconfig`` and without
   a marker file. The behavior does not change:

   * Can you ``pip install`` without an override? Currently
     yes; after this PEP yes.

   * Can it delete externally-installed packages? Because
     this is an unpatched, locally installed CPython not
     shipped through a distribution, there isn't really a
     concept of "externally-installed" packages in this use
     case; none of the ``sysconfig`` schemes of this
     interpreter should overlap any distribution-installed

>      Python you may have on the same system. Still,

Reviewer: This is slightly misunderstandable to me since an unpatched installation does not necessarily mean the absence of an external package manager operating under the same paths, only that the Python distributor (the person who `make install`-ed it) chose not to care about it. I might say something like

> By not patching the CPython installation, it is built without the considerations raised in this PEP and therefore should not be installed into a ``sysconfig`` scheme that overlaps with any distribution-installed Python on the same system.

Author: I rephrased the surrounding text in a way that hopefully makes this clearer

     version (except on Debian). After this change, the old
     version will still be on disk, but pip will
     still _shadow_ externally-installed packages, and we
     believe this to be sufficient for this not to be a
     breaking change in practice - a Python ``import``
     statement will still get you the newly-installed
     package.

     If it becomes necessary to have a way to do this, we
     suggest that the distribution should document a way for
     the installer tool to access the ``sysconfig`` scheme
     used by the distribution itself.  See the
     `Recommendations for distributions`_ section for more
     discussion.

   It is the view of the authors of this PEP that it's still
   a good idea to use virtual environments even in
   single-application container images. For instance, the
   example from `Docker's Python image`_ would be better
   implemented as::

       FROM python:3
       WORKDIR /usr/src/app
       RUN python3 -m venv venv
       RUN venv/bin/pip install -U pip
       COPY requirements.txt ./
       RUN venv/bin/pip install --no-cache-dir -r requirements.txt
       COPY . .
       CMD ["venv/bin/python", "./your-daemon-or-script.py"]

>    .. _`Docker's Python image`: https://hub.docker.com/_/python

Reviewer: The Python image is a special case here because the `python` command it references is not a distribution-provided installation, but compiled from scratch and installed into `/usr/local`. In other words, it is Case 1 instead. So this block here can be confusing.

Author: Yes, I incorrectly thought that it was Debian Python because the image was based on Debian. Someone else mentioned this too but I lost track of that. I've removed this whole example.

   This ensures that if you call out to OS-provided software
   in your code, write a downstream ``Dockerfile`` that
   applies OS upgrades, etc., any packages you've installed
   cannot impact the OS.

6. A Conda environment, where a non-``conda`` tool like pip
   is being used to remove packages. In this context, Conda
   acts as the external package manager and pip as the
   Python-specific one. This is currently supported by Conda
   as a way to install software not available in the Conda

>    repositories.

Reviewer: I am probably reading too deep into the lines, but this section makes me a bit uneasy since it feels like we’re telling Conda developers how they should do things. I feel we should just say Conda is considered a distribution and is free to choose whether to follow what is proposed in this PEP or not. (And we can add a use case for them at that time if any of the Conda folks decide to join the discussion of this PEP.)

Author: My goal was describing the impact of the PEP on Conda as it exists (as in, what happens if pip implements this PEP, does it break use of new vesions of pip with Conda?), not prescribing anything to Conda - let me know if I can make that clearer somewhere.

Author: I rephrased this to make it clearer that we're not telling Conda developers what to do, we're analyzing the behavior of this PEP and the modified pip on current Conda and making sure it's okay.

exit with an error message indicating that package
installation into this Python interpreter's directory are
disabled outside of a virtual environment.

The installer should have a way for the user to override
these rules, such as a command-line flag
``--break-system-packages``. This option should not be
enabled by default and should carry some connotation that
its use is risky.

The ``EXTERNALLY-MANAGED`` file is a metadata file in the
`packaging core metadata format`_, which is an
email-message-like format with headers and a body.  (At
current writing, that format is defined exactly as what the
standard library ``email.parser`` module can parse using
``policy=email.policy.compat32``.) If the file can be parsed
as a core metadata file, then the installer should output an
error message from that file as part of its error. If
``locale.getlocale(locale.LC_MESSAGES)`` returns
non-``None`` and the first element is a string of the form
``xx_YY``, and the file contains a header variable
``Error-xx_YY`` or failing that ``Error-xx``, then the
installer should use the value of that header as the error.
Otherwise, it should use the body of the message as an

> error.

Reviewer: What is the advantage of using the email-ish format here? The packaging community is actually not particularly satisfied with the format, and there is always an undercurrent trying to suggest something else, so I’m hesitant to make more things (not related Python package distriution) adopt it. Would an INI file not be enough?

Also, Core Metadata is referring to both the format and the field schema, so I wouldn’t use the term Core Metadata here (if we decide to go with the file format).

Author: We were going to do either TOML or "INI as parsed by ConfigParser" but @pradyunsg expressed a preference to use the email-ish format for now to avoid having to think about TOML standardization yet. Search Discord for "closer to METADATA" and you'll see the discussion.

I don't mind INI. Is it popular enough to use it? It feels like it's not used much in the packaging space, but I guess there's pydistutils.cfg and setup.cfg and pip.conf....

Anyway, I think once TOML is in the standard library we should switch. Perhaps that requires specifying that here. Maybe something like "Or, if a file `externally-managed.toml` exists, treat it like an `EXTERNALLY-MANAGED` file is present, and this PEP does not currently specify the error message behavior."

Reviewer: venv uses a subset of the INI syntax (for ease of parsing with C code without dependencies). Although `EXTERNALLY-MANAGED` probably don’t need to be parsed in environments *that* constrained, I prefer INI-like to both email-like and TOML for similar reasons. This file does not need anything past the most basic construct: simple string—even booleans are probably not needed. So I perfer keeping things as simple as possible, maybe even simpler than INI: each key-value pair is on a line, separated by `=` and stripped of `' \n\r'`.

Reviewer 5: Honestly, any of the three formats we've talked about here will work for pip, because it already has the ability to read all three of the formats. And, the specific choice doesn't really affect distro maintainers much, since I expect they'd mostly add this in via a patch.

I suggested following METADATA / PKG-INFO / WHEEL file formats because anything that needs to read this file will definitely have the reader for that format already. Otherwise, _every_ wheel installer tool will need to know how to read from TOML (tricky, with the current state of the Python TOML story) or use a specific configuration of the standard library config parser (which is less annoying nowadays, thanks to Python 2 being dead).

Reviewer 5: Ah, for that custom `key = value` style file, that is modeled after pyvenv.cfg, that works too. pip has to read that file today, and we can just reuse that code. 🙃

Basically, anything we pick here works for pip. _I_ think following the email-like format makes it so that other future wheel installation tools (you know, like the ones I'm writing) don't need to know about additional file formats.

Author: Yeah, it occurred to me that `installer` is affected here (and I snuck in a link to it :) ), and also that `setup.py` is affected - `sudo ./setup.py install` also ought to refuse to run by default. So we'll need something that both installer and setuptools are going to be happy with.

I agree that any of the formats are fine - I want to nail down the semantics of what to do about the file if it exists, and then we can deal with the detail of the file format.

Author: From the other discussion about parsing and newlines, I changed this to INI-as-parsed-by-ConfigParser, but if it's useful to constrain the syntax for non-Python parsers I'm happy to do that.

(It seems convenient to make sure the file is still parsable by ConfigParser, I think, but I'll happily defer to implementors.)

Author: I think this is still open - do we want to constrain the file format any further for the sake of non-Python parsers, or is "INI-as-parsed-by-ConfigParser" sufficient?

Reviewer: Let’s use INI-as-parsed-by-ConfigParser (I think that’s the conclusion in the call). If we add key-values in the file in the future, we should make the keys `all-lower-case` for compatibility (many INI parsers mess this up so we should be forgiving if we can).

   install`` permit installations by default, after
   implementation of this PEP?

2. If you do run such a tool, should it be willing to delete
   packages shipped by the external (non-Python-specific)
   package manager for that context, such as a distribution
   package manager?

(For simplicity, this section discusses pip as the
Python-specific installer tool, though the analysis should
apply equally to any other Python-specific package
management tool.)

This table summarizes the use cases discussed in detail
below:

==== ================================= =========================== ===================================================
Case Description                       ``pip install`` permitted   Deleting externally-installed packages permitted
==== ================================= =========================== ===================================================
1    Unpatched CPython                 Currently yes; stays yes    Currently yes; stays yes
2    Distribution ``/usr/bin/python3`` Currently yes; becomes no   Currently yes (except on Debian); becomes no
                                       (assuming the distribution

>                                        adds a marker file)

Reviewer 6: Note: "Modern" Python RPM packages on Fedora don't ship the RECORD file in their dist.info directories and hence are also not removable by `pip uninstall` or `pip install --upgrade`.

3    Distribution Python in venv       Currently yes; stays yes    There are no externally-installed packages
4    Distribution Python in venv       Currently yes; stays yes    Currently no; stays no
     with ``--system-site-packages``
5    Distribution Python in Docker     Currently yes; stays yes    Currently yes; becomes no
                                       (assuming the Docker image

>                                        removes the marker file)

Reviewer 6: ```suggestion
5    Distribution Python in Docker     Currently yes; becomes no   Currently yes; becomes no
                                       (however container images
                                        can remove the marker file)
```

Author: I think I want to lean towards expecting distributions / Docker base image providers to remove this marker file, because people are gonna get mad about not being able to `pip install` in a Docker container without a virtualenv (people already don't like the warning they now get :) ).

to how the ``PATH`` environment variable typically includes
``/usr/local/bin:/usr/bin`` and non-distribution software
installs to ``/usr/local`` by default. This split is
`recommended by the Filesystem Hierarchy Standard`__.

.. __: https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch04s09.html

There are two ways you could do this. One is, if you are
building and packaging Python libraries directly (e.g., your
packaging helpers unpack a PEP 517-built wheel or call
``setup.py install``), arrange for those tools to use a
directory that is not in a ``sysconfig`` scheme but is still
on ``sys.path``.

The other is to arrange for the default ``sysconfig`` scheme
to change when running inside a package build versus when
running on an installed system. The ``sysconfig``
customization hooks from bpo-43976_ should make this easy:
make your packaging tool set an environment variable or some
other detectable configuration, and define a
``get_preferred_schemes`` function to return a different
scheme when called from inside a package build. Then you can
use ``pip install`` as part of your distribution packaging.

.. _bpo-43976: https://bugs.python.org/issue43976

> We propose adding a ``--scheme=...`` option to instruct pip

Reviewer 6: <del>

Generally, to make things simpler for us (the distributors), I'd rather see an environment variable based approach that `sysconfig` will respect. That way:

 - upstreams don't need to implement the `--scheme=...` option in `pip`, `installer`, `setuptools`-`setup.py install` and possibly other tools
 - distributors don't need to check if all distribution packages properly use the `--scheme=...` option when invoking `python setup.py install`, `pip`, `installer` etc. (e.g. via cmake, autotools or custom makefiles)
 - distributors merely make sure that `PYTHONINSTALLSCHEME` (or similar) is set when building all distribution packages

</del>

Author: Do you think we can/should make this part of upstream sysconfig itself, or is it enough to have a recommended snippet to pass to the new option from https://bugs.python.org/issue43976 ?

Reviewer 6: Sorry, thsi particular comment was supposed to be removed before sending. It was reworked to the next one after I re-read the section.

Reviewer 6: However, since you've asked, having the snippet somewhere in the recommendations for distributors would be helpful.

to run against a specific scheme. (See `Implementation
Notes`_ below for how pip currently determines schemes.)
Once that's available, for local testing and possibly for
actual packaging, you would be able to run something like
``pip install --scheme=posix_distro`` to explicitly install
a package into your distribution's location (bypassing
``get_preferred_schemes``). One could also, if absolutely
needed, use ``pip uninstall --scheme=posix_distro`` to use
pip to remove packages from the system-managed directory,
which addresses the (hopefully theoretical) regression in
use case 5 in Rationale_.

To install packages with pip, you would also need to either
suppress the ``EXTERNALLY-MANAGED`` marker file to allow pip
to run or to override it on the command line.  You may want
to use the same means for suppressing the marker file in

> build chroots as you do in container images.

Author: @achernya pointed out to me that this is probably going to be very awkward for build chroots. (I was mostly imagining the Debian case where you can solve this with a package that dpkg-diverts the file and install that in both Docker containers and build chroots... but even there, if it's a build-dependency it's too easy to install that on a normal system.)

Here's another approach: if we extend the EXTERNALLY-MANAGED file to have a header `Permitted-Scheme: deb_system` (or whatever you name your distro scheme), then we can say that `pip install --scheme deb_system` is allowed to work without any further overrides. The semantics are that if pip is trying to install _and_ it is using any scheme other than `Permitted-Scheme`, then it's an error.

I'm not sure if we should also permit `pip install --target=/some/path` . I think the answer might be yes, so the logic would become:
* if you're installing to distutils scheme X, then look for the EXTERNALLY-MANAGED file. If you're not using a distutils scheme, allow.
* If there's a Permitted-Scheme header and it's not X, _or_ if there's no Permitted-Scheme header, error.
* Otherwise, allow.

Moreover, since pip takes command-line options as environment variables, you can just have some build tool (or your packaging) `export PIP_SCHEME=deb_system` and third-party build scripts that internally call `pip install` will do the right thing.

The downside of this approach is, of course, that it becomes very easy to install to the systemwide layout... is that a problem? I suspect it's fine; if you're specifically typing in `deb_system` you know you're doing something unusual.

Reviewer: `--target` (and `--prefix`) is basically taking the scheme’s designated path, and then swap the prefix to use the specified path, i.e.

```python
# Scheme on Linux is posix_prefix for --prefix, posix_home for --target.
paths = sysconfig.get_paths(scheme=scheme, vars={"base": new_root, "platbase": new_root})
# Install into paths["purelib"] and path["platlib"]...
```

So would it be possible to make these situations depend on the existence (and content) of `{new_root}/EXTERNALLY-MANAGED` instead?

Reviewer 3: Yeah, a Debian package build (and presumably other distro builds) will usually build into a different `DESTDIR` (not quite prefix, in the traditional sense, because the software must still assume that `/usr` is its prefix in the final install).

For `setuptools` builds, we use `setup.py --root temporary/install/location`.
For `pip` (which we don't use in package builds, atm), I'd imagine we'd use `--root` too.

So, if purelib/platlib != the standard purelib/platlib for the default scheme, it's probably safe to install to.

@geofft: Yes, I agree that explicitly selecting `--scheme deb_system` should imply that you want to clobber things in `deb_system`

Author: Hmm so I wrote up https://github.com/geofft/peps/pull/6 but I missed the comment above about how `--target` and `--prefix` do involve looking up the default scheme and modifying it.

On my Debian stable box, `base` doesn't cause `stdlib` to change:
```
$ python3
Python 3.7.3 (default, Jul 25 2020, 13:03:44) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sysconfig
>>> sysconfig.get_paths(vars=dict(base="foo"))
{'stdlib': '/usr/lib/python3.7', 'platstdlib': '/usr/lib/python3.7', 'purelib': 'foo/lib/python3.7/site-packages', 'platlib': '/usr/lib/python3.7/site-packages', 'include': '/usr/include/python3.7m', 'platinclude': '/usr/include/python3.7m', 'scripts': 'foo/bin', 'data': 'foo'}
```

(It does on Python 2, amusingly enough.)

So making this depend on whether the final `stdlib` path contains an `EXTERNALLY-MANAGED` file won't work right - it still will.

Should we explicitly say "and if you are not overriding variables in the scheme to move things around"? I'm not quite sure how to phrase this.

I suppose the answer is some phrasing to the effect of "If you're using an _unmodified_ distutils scheme" but also phrasing it in a way where `--root` doesn't count as modifying it. That is, even though `pip install --root $DESTDIR` installs to `$DESTDIR/usr/local/lib/python3.x/site-packages` instead of `/usr/local/lib/python3.x/site-packages`, it should still get you an error. (As a concrete example, there's no real difference between an FPM-packaged thing in /usr/local/lib and a directly-installed-by-the-sysadmin thing. If you're going to be installing Python packages with FPM, you still need to put them in a virtualenv, and maybe you want `dh_virtualenv`.)

Reviewer 3: > On my Debian stable box, base doesn't cause stdlib to change:

Because it uses `installed_base`:
```py
    'posix_prefix': {
        'stdlib': '{installed_base}/{platlibdir}/python{py_version_short}',
        'platstdlib': '{platbase}/{platlibdir}/python{py_version_short}',
        'purelib': '{base}/lib/python{py_version_short}/site-packages',
        'platlib': '{platbase}/{platlibdir}/python{py_version_short}/site-packages',
        'include':
            '{installed_base}/include/python{py_version_short}{abiflags}',
        'platinclude':
            '{installed_platbase}/include/python{py_version_short}{abiflags}',
        'scripts': '{base}/bin',
        'data': '{base}',
        },
```

> and if you are not overriding variables in the scheme to move things around"?

Yeah, the goal is to be sure that we're installing off of the path that a default system python will use.

But it's hard to know what those `sys.path` components will be. Debian's `site.py` inserts some things, probably more than other distros do. I can't think of a foolproof way to know what values are, without firing up a child `/usr/bin/python` in a controlled environment (no `PYTHONPATH`, `PYTHONHOME`, etc.). Although even that could miss directories that would be on `sys.path` if they existed. Maybe explicitly listing them in EXTERNALLY-MANAGED is the only way?

Reviewer: Yes, sorry, my code example above was only illustrative and not correct, because I didn’t want the giant `base` `installed_base` `platbase` etc. dict to obscure the point I wanted to make. I should’ve pointed this out in the comment.

The advantage of setting these up to be automatic
(suppressing the marker file in your build environment and
having ``get_preferred_schemes`` automatically return your
distribution's scheme) is that an unadorned ``pip install``
will work inside a package build, which generally means that
an unmodified upstream build script that happens to
internally call ``pip install`` will do the right thing.
You can, of course, just ensure that your packaging process
always calls ``pip install --scheme=posix_distro

> --break-system-packages``, which would work too.

Reviewer 6: Also, it makes things simpler for us (the distributors):

 - upstreams don't need to implement the `--scheme=...` option in `pip`, `installer`, `setuptools`-`setup.py install` and possibly other tools right away
 - distributors don't need to check if all distribution packages properly use the `--scheme=...` option when invoking `python setup.py install`, `pip`, `installer` etc. (e.g. via cmake, autotools or custom Makefiles)
 - distributors don't need to create backwards compatibility shims for their versions where `--scheme=...`  is not yet available/used

Reviewer: pip is probably going to implement `--scheme` regardless, but the second point is convincing to me. This way distributors would control both the scheme selection logic and the schemes, removing a possibility of miscommunications. It also means distributors can iterate faster implementing and changing the schemes if needed.

This also means less work for pip (and other installers), but I’d say it’s incidental 🙂

   should not install a custom CPython built with
   ``./configure --prefix=/usr``, or it will overwrite some
   files from the distribution and the distribution will
   eventually overwrite some files from your installation.
   Instead, your installation should be in a separate
   directory (perhaps ``/usr/local``, ``/opt``, or your home
   directory).

   Therefore, we can assume that such a CPython has its own
   ``stdlib`` directory and its own ``sysconfig`` schemes
   that do not overlap any distribution-installed Python. So
   any OS-installed packages are not visible or relevant
   here.

   If there is a concept of "externally-installed" packages
   in this case, it's something outside the OS and generally
   managed by whoever built and installed this CPython.
   Because the installer chose not to add a marker file or
   modify ``sysconfig`` schemes, they're choosing the
   current behavior, and ``pip install`` can remove any
   packages available in this CPython.

2. A distribution's ``/usr/bin/python3``, either when
   running ``pip install`` as root or ``pip install
   --user``, following our `Recommendations for

>    distributions`_.

Reviewer 4: Question: is there a way to limit `pip install` while still allowing `pip install --user`? I've read the whole PEP (in 1-2 sittings) and didn't see anything explicit about it.

My main context is the Windows Store package of the python.org distro, which has a very locked down system site-packages (containing packages that would otherwise be upgradeable), but also sumo distros in general (which I guess is kind of similar to the monorepo case). If I've got a standard distro including packages that are preferred, but still want to allow individual users to override (not remove/replace) them, I'd love to have the mechanism be standardised.

I'm guessing bundling the user packages in with site packages is because none of the distros currently use the `-s` option when invoking their own commands?

Reviewer 3: `pip install --user` can still cause these issues. Here's an example bug I was looking at, today, caused by a user not realising what they'd installed to `--user`: https://bugs.launchpad.net/bugs/1912535

> My main context is the Windows Store package of the python.org distro

I wouldn't expect that to set EXTERNALLY-MANAGED, because they don't provide Python libraries in the Windows Store (do they?). So, they shouldn't be concerned about pip installing things. Even if file permissions mean that `--user` is the only way pip can install, that shouldn't be an issue.

Reviewer 4: > Here's an example bug I was looking at, today, caused by a user not realising what they'd installed to --user

Or arguably, caused a system-installed tool _not_ passing `-s` to ignore user-installed packages?

> I wouldn't expect that to set EXTERNALLY-MANAGED, because they don't provide Python libraries in the Windows Store (do they?)

We could, though the amount of overhead isn't really worth it (at least at the ridiculously high rate of releases everything has). And yeah, the super-restrictive permissions are going to break anyone who tries to install into the default directory anyway. Pip already has fallback behaviour, but everyone else will figure it out eventually after they get bug reports (unless there's a spec that they're already following to detect that they shouldn't even attempt to use that folder and to display a message explaining why...)

Seeing as the `EXTERNALLY-MANAGED` file can already contain options, why not have an option for `managed-schemes: posix_user, nt_user` or similar? Make it default to what's already specified in the PEP, and then it can be overridden for fewer/greater restrictions?

Reviewer 3: > Or arguably, caused a system-installed tool not passing -s to ignore user-installed packages?

Nothing specifies `-s`. Or at least, I've never seen anything that does.

Reviewer 4: Petr posted on linux-sig that Red Hat is moving to start using `-s` everywhere that they want to avoid conflicts with user-specified packages. So apparently it's possible for distros to change here, which would enable user packages to be used as intended.

Author: Right - if the goal here is "pip shouldn't install systemwide because the user doesn't have permissions" as opposed to "pip shouldn't install systemwide because it will risk breaking software shipped expecting a particular Python," then I think the thing most in line with the worldview of this PEP is to set your default sysconfig scheme to be `nt_user` or something, so that that's where `pip install` installs things to, and don't configure anything else. Don't set `EXTERNALLY-MANAGED`. You can still ship software with that installation of Python in the `nt` scheme and it will get loaded, and because of the second change of this PEP, pip won't remove them if it installs a new version of something to `nt_user`.

(I actually think it already won't remove them, but I'm not positive. Also note that I haven't tested that setting the default sysconfig scheme to a user scheme causes pip to do what you expect, but it seems like it ought to / this PEP sort of assumes that installers use the default sysconfig scheme by default.)

`EXTERNALLY-MANAGED` is for when you're simultaneously shipping Python as a thing for end users to use directly and also as an implementation language for software that you want to not be fragile. If I understand your use case correctly, there's no software using the Windows Store distribution of Python as an interpreter (other than Python-specific stuff like pip, idle, pydoc, etc.), right?

Or put another way, I think your goal is not to prevent users adding things to the global (virtualenv-less) `sys.path` and breaking things, because there's nothing they would break; your goal is simply to prevent installers from defaulting to the global installation directory.

But I don't know a ton about this use case and might be misunderstanding it.

None of the distros currently use `-s` when invoking their own commands because there are legitimate reasons to add software to the global context (e.g., running /usr/bin/ansible but installing Ansible extensions / writing custom Python). The position of this PEP is that adding software to the global context should take explicit user action but it should, generally, be permitted (or at least that distros should be able to opt particular applications into supporting it).

Reviewer 4: I guess I'm still not getting how you can allow users to install with `--user` by preventing them from installing with `--user`. Is the plan to teach everyone to just add another option to be able to install things "globally"? Because that doesn't sound like much of an improvement over the current situation.

> or at least that distros should be able to opt particular applications into supporting it

The officially supported upstream way of doing this on a per-application level is the `-s` option. So you would default every application to use `-s`, and then those that want to use user-installed extensions should omit it. Adding (or removing) the user site packages path directly is also fairly straightforward if you control the application (and e.g. want a CLI option for "safe mode" that ignores user packages).

Basically what I'm getting at is that I totally agree with an interpreter-level instruction to package managers "you should not be modifying this directory", but I think we're pushing it too far and forcing legitimate, supported scenarios into needing workarounds. We don't have to push it too far.

Reviewer 3: > I guess I'm still not getting how you can allow users to install with --user by preventing them from installing with --user.

The goal is to make them jump through a hoop if they want to install with `--user`. Don't do it by default, but make them explicitly opt-in. And realise what they're opting into, e.g. suggest alternatives that are better for what they're trying to do.

`sudo` was not a sufficient opt-in for doing things globally that can break your system, because people are used to sudo enough that they just do it without thinking.

Reviewer 4: Why won't that happen again with the new flag? Are we relying on Python being enough of an edge case that people won't internalise (or write blog posts about) the new way?

Reviewer 3: Because we'll be suggesting mechanisms that actually work, e.g. pipx and virtualenvs, rather than just saying if you really want to do this, use a bigger hammer.

Reviewer 4: Those aren't going to work for adding an extension to system-installed Ansible though, right? (to pick up on the example given above)

Reviewer 3: Right, that's the situation where you need to pass a flag to say, yes I really mean to do this.

installer should ignore the parse failure and instead just
use a pre-defined error message of its own, which should
suggest that the user create a virtual environment to
install packages.

Software distributors who have a non-Python-specific package
manager that manages libraries in the ``sys.path`` of their
Python package should, in general, ship a
``EXTERNALLY-MANAGED`` file in their standard library
directory. For instance, Debian may ship a file in
``/usr/lib/python3.9/EXTERNALLY-MANAGED`` consisting of
something like

::

    To install Python packages systemwide, try apt install
    python3-xyz, where xyz is the package you are trying to
    install.

    If you wish to install a non-Debian-packaged Python
    package, create a virtual environment using python3 -m
    venv path/to/venv. Then use path/to/venv/bin/python and
    path/to/venv/bin/pip. Make sure you have python3-full
    installed.

    If you wish to install a non-Debian packaged Python
    application, it may be easiest to use pipx install xyz,
    which will manage a virtual environment for you. Make
    sure you have pipx installed.

    See /usr/share/doc/python3.9/README.venv for more

>     information.

Reviewer 3: This does not parse as an `EmailMessage`, but as a `Message`. By the rule above rule, this text would then be ignored.

It would probably be good to use an example with headers to show how paragraphs are handled in headers.

Reviewer 5: FWIW, the format is gonna be to match METADATA files.

That's a bit underspecified (see start of https://packaging.python.org/specifications/core-metadata/). Basically everything uses the exact same code for this, which looks like: https://installer.readthedocs.io/en/latest/api/utils/#installer.utils.parse_metadata_file

Author: Hm. I think what's going on is that `email.parser.Parser()` defaults in all current versions of Python to `policy=email.policy.compat32`, which always returns a `Message` - if you change the policy to `email.policy.default`, then it returns an `EmailMessage`.

Also, the source code for `parse_metadata_file` is just

```python
def parse_metadata_file(contents):
    # type: (Text) -> Message
    """Parse :pep:`376` ``PKG-INFO``-style metadata files.

    ``METADATA`` and ``WHEEL`` files (as per :pep:`427`) use the same syntax
    and can also be parsed using this function.

    :param contents: The entire contents of the file.
    """
    feed_parser = FeedParser()
    feed_parser.feed(contents)
    return feed_parser.close()
```

which I think is equivalent to the code I have (just using feed vs. non-feed mode).

I suspect both of us need to explicitly specify `email.policy.compat32`, because the docs say the default will change in the future.

Also, I'm trying to figure out under what circumstances something _doesn't_ parse (using `compat32`). It looks like there's a concept of "defects" which are added to `m.defects`, but they don't raise, and the example above causes a `MissingHeaderBodySeparatorDefect`, but you can ignore it. I'm starting to suspect the answer is that every possible string parses with `compat32`.

(Curiously, `From: "` doesn't cause `email.policy.default`  to throw on parsing but to get you an `EmailMessage` whose `items()` method throws, at least in Python 3.7.  With `compat32`, `From: "` gets you a `Message` whose `items()` is `[('From', '"')]`.)

Author: I'm changing this to link to https://packaging.python.org/specifications/core-metadata/ (as PEP 566 does), and mentioning that this happens to be defined as stdlib `email.parser` with `policy=compat32`. I'm worried that an example with headers will be too long here and disrupt the flow of reading, and I think deferring to the core metadata spec etc. will explain how to implement headers with paragraphs.

I'm also a bit confused in that https://packaging.python.org/specifications/core-metadata/#description says that you can use seven spaces and a pipe character to continue long paragraphs, but,
```python
>>> m = installer.utils.parse_metadata_file("""Description: This project provides powerful math functions
...         |For example, you can use `sum()` to sum numbers:
...         |
...         |Example::
...         |
...         |    >>> sum(1, 2)
...         |    3
...         |""")
>>> m["Description"]
'This project provides powerful math functions\n        |For example, you can use `sum()` to sum numbers:\n        |\n        |Example::\n        |\n        |    >>> sum(1, 2)\n        |    3\n        |'
```

Is this behavior specific to the interpretation of the Description field and not part of the document format in general?

Reviewer: The long string logic of the metadata format is *extremely* underspecified and buggy, which is one of the biggest reasons I strongly dislike using that format elsewhere. Almost no metadata files in the real world actually uses that, and when they do it's an exercise of trail and error figuring out what is actually being converted to what.

Author: OK, that's as good a reason as any to decide against that format, since `configparser` is [very clear about multi-line handling](https://docs.python.org/3/library/configparser.html#supported-ini-file-structure):

```python
>>> import configparser
>>> c = configparser.ConfigParser()
>>> c.read_string("[externally-managed]\nError: pip install is prohibited outside virtualenvs\n\n Virtualenvs are cool")
>>> c["externally-managed"]["error"]
'pip install is prohibited outside virtualenvs\n\nVirtualenvs are cool'
```

I'll make this update and I'll also try to come up with a multi-language example.

  error message indicating both how to install system-wide
  packages via the distribution's package manager and how to
  set up a virtual environment. If your distribution is
  often used by users in a state where the ``python3``
  command is available (and especially where ``pip`` or
  ``get-pip`` is available) but ``python3 -m venv`` does not
  work, the message should indicate clearly how to make
  ``python3 -m venv`` work properly.

* Consider packaging pipx_, a tool for installing
  Python-language applications, and suggesting it in the
  error. It automatically creates a virtual environment for
  that application alone, which is a much better default for
  end users who want to install some Python-language
  software (which isn't available in the distribution) but
  are not themselves Python users. Packaging pipx in the
  distribution avoids the irony of instructing users to
  ``pip install --user --break-system-packages pipx`` to
  _avoid_ breaking system packages.  Consider arranging
  things so your distribution's package / environment for
  Python when used directly (e.g., ``python3`` on Fedora or
  ``python3-full`` on Debian) depends on pipx in addition to

>   pip.

Reviewer 3: `python3-full` doesn't depend on `pip`. It does depend on `ensurepip`, so that `venv` works, but you only get `pip` inside the virtualenv, not outside it.

  system interpreter's  ``sys.path``, one for
  distribution-installed packages and one for packages
  installed by the local system administrator, and configure
  ``sysconfig.get_default_scheme()`` to point at the latter
  path. This ensures that tools like pip will not modify
  distribution-installed packages. The path for the local
  system administrator should come before the distribution
  path on ``sys.path`` so that local installs take
  preference over distribution packages.

  For example, Fedora and Debian (and their derivatives)
  both implement this split by using ``/usr/local`` for
  locally-installed packages and ``/usr`` for
  distribution-installed packages. Fedora uses
  ``/usr/local/lib/python3.x/site-packages`` vs.
  ``/usr/lib/python3.x/site-packages``. Debian uses
  ``/usr/local/lib/python3.x/dist-packages`` vs.
  ``/usr/lib/python3.x/dist-packages`` as an additional
  layer of separation from a locally-compiled Python
  interpreter: if you build and install upstream CPython in
  ``/usr/local/bin``, it will look at
  ``/usr/local/lib/python3.x/site-packages``, and Debian
  wishes to make sure that packages installed via

>   ``/usr/local/bin/pip`` for the locally-built interpreter

Reviewer 3: `/usr/local/bin/pip` is ambiguous. It could either be a manually installed pip in the distro-python or in the locally-compiled python. It may be clearer to avoid mentioning pip here, and just talk about packages installed into the locally-compiled python being on a separate sys.path to packages built for the distro python.
